xdp/src/device.rs

```diff
@@ -0,0 +1,495 @@
+        let fd = unsafe { libc::socket(libc::AF_INET, libc::SOCK_DGRAM, 0) };
+        if fd < 0 {
+            return Err(io::Error::last_os_error());
+        }
+        let fd = unsafe { OwnedFd::from_raw_fd(fd) };
+
+        let mut req: ifreq = unsafe { mem::zeroed() };
+        let if_name = CString::new(self.if_name.as_bytes()).unwrap();
+
+        let if_name_bytes = if_name.as_bytes_with_nul();
+        let len = std::cmp::min(if_name_bytes.len(), IF_NAMESIZE);
+        unsafe {
+            std::ptr::copy_nonoverlapping(
+                if_name_bytes.as_ptr() as *const c_char,
+                req.ifr_name.as_mut_ptr(),
+                len,
+            );
+        }
+
+        let result = unsafe { ioctl(fd.as_raw_fd(), SIOCGIFHWADDR, &mut req) };
+        if result < 0 {
+            return Err(io::Error::last_os_error());
+        }
+
+        Ok(unsafe {
+            slice::from_raw_parts(req.ifr_ifru.ifru_hwaddr.sa_data.as_ptr() as *const u8, 6)
+        }
+        .try_into()
+        .unwrap())
+    }
+
+    pub fn ipv4_addr(&self) -> Result<Ipv4Addr, io::Error> {
+        let fd = unsafe { libc::socket(libc::AF_INET, libc::SOCK_DGRAM, 0) };
+        if fd < 0 {
+            return Err(io::Error::last_os_error());
+        }
+        let fd = unsafe { OwnedFd::from_raw_fd(fd) };
+
+        let mut req: ifreq = unsafe { mem::zeroed() };
+        let if_name = CString::new(self.if_name.as_bytes()).unwrap();
+
+        let if_name_bytes = if_name.as_bytes_with_nul();
+        let len = std::cmp::min(if_name_bytes.len(), IF_NAMESIZE);
+        unsafe {
+            std::ptr::copy_nonoverlapping(
+                if_name_bytes.as_ptr() as *const c_char,
+                req.ifr_name.as_mut_ptr(),
+                len,
+            );
+        }
+
+        let result = unsafe { libc::ioctl(fd.as_raw_fd(), SIOCGIFADDR, &mut req) };
+        if result < 0 {
+            return Err(io::Error::last_os_error());
+        }
+
+        let addr = unsafe {
+            let addr_ptr = &req.ifr_ifru.ifru_addr as *const libc::sockaddr;
+            let sin_addr = (*(addr_ptr as *const libc::sockaddr_in)).sin_addr;
+            Ipv4Addr::from(sin_addr.s_addr.to_ne_bytes())
+        };
+        Ok(addr)
+    }
+
+    pub fn open_queue(&self, queue_id: QueueId) -> DeviceQueue {
+        DeviceQueue::new(self.if_index, queue_id)
+    }
+}
+
+pub struct DeviceQueue {
+    if_index: u32,
+    queue_id: QueueId,
+    completion: Option<TxCompletionRing>,
+}
+
+impl DeviceQueue {
+    pub fn new(if_index: u32, queue_id: QueueId) -> Self {
+        Self {
+            if_index,
+            queue_id,
+            completion: None,
+        }
+    }
+
+    pub fn if_index(&self) -> u32 {
+        self.if_index
+    }
+
+    pub fn id(&self) -> QueueId {
+        self.queue_id
+    }
+
+    pub fn tx_completion(&mut self) -> Option<&TxCompletionRing> {
+        self.completion.as_ref()
+    }
+}
+
+pub(crate) struct RingConsumer {
```

Comment:
This has enough unsafe to justify its own file and plenty of tests to make sure we are not harboring footguns here.

Reply:
RingProducer and RingConsumer are spsc. There are tests. Which tests do you think are missing?

Reply:
1 producer thread, 1 consumer thread, 3 setups: roughly equal prod and cons rates, one where producer is much faster, one where consumer is much faster. Go over some 1000 loops of the ringbuffer in both cases. If producer pushes consequitive integers, you know exactly the sum the receiver should be getting to have received them all.

Reply:
I think you mean use loom

Reply:
loom would be great but we are not shipping crossbeam here as you have correctly pointed out=) Either way, not super critical as long as this is not ingress.